pkg/patterns/application/application.go

```diff
@@ -0,0 +1,153 @@
+	return reconcile.Result{}, nil
+})
+
+var getConfig = config.GetConfig
+var newController = controller.New
+var newManager = manager.New
+var getGvk = apiutil.GVKForObject
+
+// Builder builds an Application Controller and returns the manager.Manager to start the Controller.
+type Builder struct {
+	apiType runtime.Object
+	mrg     manager.Manager
+	pred    []predicate.Predicate
+	manages []runtime.Object
+	config  *rest.Config
+	r       reconcile.Reconciler
+}
+
+// BuilderFor returns a new Builder for and BuilderFor
+func BuilderFor(apiType runtime.Object) *Builder {
+	return &Builder{apiType: apiType}
+}
+
+// Owns configures the Application Controller to respond to create / delete / update events for objects it manages
+// - e.g. creates.  apiType is an empty instance of an object matching the managed object type.
+func (b *Builder) Owns(apiType runtime.Object) *Builder {
+	b.manages = append(b.manages, apiType)
+	return b
+}
+
+// WithConfig sets the Config to use for configuring clients.  Defaults to the in-cluster config or to ~/.kube/config.
+func (b *Builder) WithConfig(config *rest.Config) *Builder {
+	b.config = config
+	return b
+}
+
+// WithManager sets the Manager to use for registering the Controller.  Defaults to a new manager.Manager.
+func (b *Builder) WithManager(m manager.Manager) *Builder {
+	b.mrg = m
+	return b
+}
+
+// WithEventFilter sets the event filters.  Defaults to the empty list.
+func (b *Builder) WithEventFilter(p predicate.Predicate) *Builder {
+	b.pred = append(b.pred, p)
+	return b
+}
+
+// WithHandler sets th Reconcile called in response to create / update / delete events for the
+// BuilderFor type or Created object types.
+func (b *Builder) WithHandler(r reconcile.Reconciler) *Builder {
+	b.r = r
+	return b
+}
+
+// Build builds the Application Controller and returns the Manager used to start it.
+func (b *Builder) Build() (manager.Manager, error) {
+	a := &application{}
+
+	if b.r == nil {
+		return nil, fmt.Errorf("must call Builder.WithHandler to set Reconcile")
+	}
+
+	var err error
+	if b.config == nil {
+		b.config, err = getConfig()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if b.mrg == nil {
+		b.mrg, err = newManager(b.config, manager.Options{})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	a.mrg = b.mrg
+
+	gvk, err := getGvk(b.apiType, a.mrg.GetScheme())
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the controller
+	name := fmt.Sprintf("%s-application", strings.ToLower(gvk.Kind))
+	ctrl, err := newController(name, a.mrg, controller.Options{Reconciler: b.r})
+	if err != nil {
+		return nil, err
+	}
+	a.ctrl = ctrl.(*internalcontroller.Controller)
+
+	// Reconcile type
+	err = a.ctrl.Watch(&source.Kind{Type: b.apiType}, &handler.EnqueueRequestForObject{}, b.pred...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Watch the managed types
+	for _, t := range b.manages {
+		err = a.ctrl.Watch(&source.Kind{Type: t}, &handler.EnqueueRequestForOwner{
```

Comment:
is there/should there be a way to add non-owns relationships with this builder?

Reply:
Not right now.  The point of this package is to offer a simplified view for the most common way to build application controllers.  If arbitrary mappings become common for operators, then we can add them.  Otherwise users can break out and use the underlying libraries directly.